### ISSUE TITLE
Compute metrics for generation tasks in SFTTrainer

### ISSUE THREAD

**Issue by wei-ann-Github** [huggingface/trl#862]
Hi, I want to include a custom generation based compute_metrics e.g., BLEU, to the `SFTTrainer`. However, I have difficulties because:
1. The input, `eval_preds`, into compute_metrics contains a `.predictions` attribute but its value are logits rather than generations. There don't seem to be an attribute for generations.
2. In `Seq2seqTrainingArguments`, there is a `predict_with_generate` argument for the purpose of generation based metrics. But this does not work on the `SFTTrainer`. So I am not able to compute the metrics that I want.

Can anyone advise how I can get the generations into the compute_metrics function, please? The `trl` version I am using is `trl==0.5.0`

Here are my TrainingArgs and SFTTrainer.
```
training_args = Seq2SeqTrainingArguments(
    output_dir=output_dir,
    eval_steps=eval_steps,
    evaluation_strategy=evaluation_strategy,
    gradient_accumulation_steps=gradient_accumulation_steps,
    learning_rate=lr,
    max_steps=max_steps,
    num_train_epochs=num_train_epochs,
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    save_steps=save_steps,
    save_strategy=save_strategy,
    warmup_steps=num_warmup_steps,
    weight_decay=weight_decay,
    save_total_limit=save_total_limit,
    load_best_model_at_end=True,
    metric_for_best_model=metric_for_best_model,
    greater_is_better=metric_greater_is_better,
    generation_config=generation_config,
    predict_with_generate=True,
    generation_max_length=10,
)

trainer = SFTTrainer(
    model,
    args=training_args,
    train_dataset=dataset_train,
    eval_dataset=dataset_eval,
    formatting_func=formatting_prompts_func,
    data_collator=collator,
    compute_metrics=compute_metrics,
)
```

Attributes of eval_preds are:
```
>> dir(eval_preds)
['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'inputs',
 'label_ids',
 'predictions']
```

Thank you in advance!

**Comment by lvwerra** [comment]
Indeed the `SFTTrainer` currently does not support `predict_with_generate` but that's a feature we could add in the future. @younesbelkada what do you think?

**Comment by ykemiche** [comment]
@lvwerra does this mean that we can't customize compute_metrics function for tracking metrics like "ROUGE"  over steps?

**Comment by lvwerra** [comment]
Not at this point. As pointed out it would require a `predict_with_generate` method which we don't have, yet.

**Comment by therealTT** [comment]
@lvwerra 
> does this mean that we can't customize compute_metrics function for tracking metrics like "ROUGE" over steps?

When you say customize compute_metrics what do you mean? Like isn't the compute_metrics already a custom function? I am also trying to track the ROUGE metric during my evaluation step, but I am running into CUDA memory issues. I think the memory issue s not related to predict_with_generate, but I'm hoping you might know what is causing it. I am using an L4 24GB GPU. 

This is my compute metrics function:

def compute_metrics(eval_preds):
    preds, labels = eval_preds
    # decode preds and labels
    decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True, device="cpu")
    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True, device="cpu")
    # Compute rouge scores
    rouge_output = rouge.compute(predictions=decoded_preds, references=decoded_labels)
    return rouge_output

I further tried to use sub batche sizes also:
def compute_metrics(eval_preds):
    preds, labels = eval_preds
    # Calculate batch size and desired sub-batch size
    batch_size = preds.shape[0]
    sub_batch_size = 3  # Adjust this based on your memory availability and desired efficiency

    # Loop through sub-batches
    decoded_preds, decoded_labels = [], []
    for i in range(0, batch_size, sub_batch_size):
        sub_preds = preds[i:min(i + sub_batch_size, batch_size)]
        sub_labels = labels[i:min(i + sub_batch_size, batch_size)]
        # Decode the sub-batch using your current decoding logic
        sub_decoded_preds = tokenizer.batch_decode(sub_preds, skip_special_tokens=True)
        sub_decoded_labels = tokenizer.batch_decode(sub_labels, skip_special_tokens=True)
        # Append the decoded sub-batches to the main lists
        decoded_preds.extend(sub_decoded_preds)
        decoded_labels.extend(sub_decoded_labels)

    # Calculate ROUGE scores using the accumulated decoded data
    rouge_output = rouge.compute(predictions=decoded_preds, references=decoded_labels)
    return rouge_output

Thanks!

**Comment by isspek** [comment]
I have same question, is there any update on this issue? Currently I can not use custom metric which is crucial to use early stopping callback in our use case.

I receive error by tokenizer.batch_decode from the compute_metric when I apply instruction-tuning on Llama2 with SFT. However, this method works without any problem with the standard Trainer.

```python
        def compute_metrics(eval_preds):
            preds, labels = eval_preds

            if isinstance(preds, tuple):
                preds = preds[0]

            # Replace -100 in the preds as we can't decode them
            preds = np.where(preds != -100, preds, self.tokenizer.pad_token_id)

            # Decode generated summaries into text
            decoded_preds = self.tokenizer.batch_decode(preds, skip_special_tokens=True)

            # Replace -100 in the labels as we can't decode them
            labels = np.where(labels != -100, labels, self.tokenizer.pad_token_id)
            # Decode reference summaries into text
            decoded_labels = self.tokenizer.batch_decode(labels, skip_special_tokens=True)
            # ROUGE expects a newline after each sentence
            decoded_preds = ["\n".join(pred.strip()) for pred in decoded_preds]

            decoded_labels = ["\n".join(label.strip()) for label in decoded_labels]
            # Compute ROUGscores
            result = rouge_score.compute(
                predictions=decoded_preds, references=decoded_labels, use_stemmer=True
            )
            # Extract the median scores
            result = {key: value * 100 for key, value in result.items()}
            return {k: round(v, 4) for k, v in result.items()}
```
Thanks in advance.

**Comment by lvwerra** [comment]
Note that the `preds` only contain the probs for the next token, so you can't really decode them into a text. We'll work on adding text generation metrics similar to the `Seq2SeqTrainer` in `transformers` soon. In the meantime if you'd like you could also have a stab at it, happy to help guide a PR on this.

**Comment by isspek** [comment]
Hi @lvwerra 
Thanks a lot for your reply. 

I have came across` preprocess_logits_for_metrics` in SFTTrainer. After changing sequential training arguments to standard one and define a function for preprocess_logits_for_metrics. I was able to use the custom metric.

Neverthless, I would be glad to use new update whenever is ready.

**Comment by therealTT** [comment]
@lvwerra

I would love to contribute to this repo. This is going to be my first contribution ever, so any guidance would be tremendously helpful. I have a faint idea of where to start, is there a preferred approach or method? Thanks!

**Comment by lvwerra** [comment]
I think looking at the `Seq2SeqTrainer` in `transformers` would be a good start, which offers that option and we also inherit from the `Trainer`. Then you could open draft PR, even if things are not fully working/ready yet and we can directly discuss on the code.

**Comment by anandsarth** [comment]
@isspek can you share your code please, I also want to compute metrics using the generations

**Comment by isspek** [comment]
Hi @anandsarth,

I used the compute metric as the following example:

```python

        def preprocess_logits_for_metrics(logits, labels):
            if isinstance(logits, tuple):
                logits = logits[0]
            return logits.argmax(dim=-1)


        def compute_metrics(eval_preds):
            preds, labels = eval_preds

            if isinstance(preds, tuple):
                preds = preds[0]

            # Replace -100 in the preds as we can't decode them
            preds = np.where(preds != -100, preds, self.tokenizer.pad_token_id)

            # Decode generated summaries into text
            decoded_preds = self.tokenizer.batch_decode(preds, skip_special_tokens=True)

            # Replace -100 in the labels as we can't decode them
            labels = np.where(labels != -100, labels, self.tokenizer.pad_token_id)
            # Decode reference summaries into text
            decoded_labels = self.tokenizer.batch_decode(labels, skip_special_tokens=True)
            # ROUGE expects a newline after each sentence
            decoded_preds = ["\n".join(pred.strip()) for pred in decoded_preds]

            decoded_labels = ["\n".join(label.strip()) for label in decoded_labels]
            # Compute ROUGscores
            result = rouge_score.compute(
                predictions=decoded_preds, references=decoded_labels, use_stemmer=True
            )
            # Extract the median scores
            result = {key: value * 100 for key, value in result.items()}
            return {k: round(v, 4) for k, v in result.items()}


        # declare training arguments
        training_args = TrainingArguments(
            output_dir=params['model_output_path'],
            evaluation_strategy="steps",
            do_eval=True,
            save_strategy="steps",
            eval_steps=100,
            learning_rate=learning_rate,
            logging_strategy='steps', # log according to log_steps
            warmup_steps=int(len(train_ds) / 8),
            weight_decay=0.01,
            save_total_limit=1,
            num_train_epochs=epochs,
            auto_find_batch_size=True,
            metric_for_best_model=params['eval_metric'],
            greater_is_better=True,
            # metric_for_best_model='eval_loss',
            load_best_model_at_end=True
        )

        # declare trainer
        trainer = SFTTrainer(
            model=model,
            args=training_args,
            train_dataset=train_ds,
            eval_dataset=val_ds,
            dataset_text_field="instruction",
            peft_config=lora_config,
            tokenizer=self.tokenizer,
            max_seq_length= max_length,
            compute_metrics=compute_metrics,
            preprocess_logits_for_metrics = preprocess_logits_for_metrics,
            callbacks=[EarlyStoppingCallback(early_stopping_patience=10)]
        )

```

**Comment by knitzschke** [comment]
Hi @isspek. I am new to transformers and huggingface. How did you implement the solution? Did you put that directly in script or in trl.trainer.sft_trainer?

Any help is appreciated!!

**Comment by knitzschke** [comment]
@isspek for implementation is you input the two functions into transformers.trainer ?

**Comment by antonioalegria** [comment]
Would changing SFTTrainer to inherit from Seq2SeqTrainer instead of Trainer be a good starting point? I notice that there is no overlap between the overridden methods in both these classes.

**Comment by NirmalManoj** [comment]
@lvwerra Why is this issue closed? Is there a way around this currently?

**Comment by NirmalManoj** [comment]
Hi @isspek, could you explain your approach in a bit more detail? What is the format of your input? It appears to me that you are not separating the prompt from the generation when calculating the scores. Is that being handled separately in your code?

**Comment by snassimr** [comment]
Any solution in SFTTrainer ? @isspek , your way to treat prompts can be very useful . Thank in advance

**Comment by pierpierpy** [comment]
i'm trying to decode with the tokenizer but it spits out some gibberish. half complete words and sentences. they make a bit of sense but they are pretty garbage. is there any solution to the problem? i'd like to have the real predictions to compute the rouge metric during training at each epoch.

**Comment by snassimr** [comment]
> i'm trying to decode with the tokenizer but it spits out some gibberish. half complete words and sentences. they make a bit of sense but they are pretty garbage. is there any solution to the problem? i'd like to have the real predictions to compute the rouge metric during training at each epoch.

Are you sure it's gibberish ? In the beginning it can look gibberish but improving .

**Comment by pierpierpy** [comment]
> > i'm trying to decode with the tokenizer but it spits out some gibberish. half complete words and sentences. they make a bit of sense but they are pretty garbage. is there any solution to the problem? i'd like to have the real predictions to compute the rouge metric during training at each epoch.
> 
> Are you sure it's gibberish ? In the beginning it can look gibberish but improving .

The phrases produced during the training are "readable" but the words are misspelled. On the other hand, the predictions made using the model after the training seems to be just right  with no grammatical mistakes or misspellings. It is strange because i also tried to make the prediction during and after training with just one epoch. I assumed they should have been both wrong, but the results after training are still good, while the results during training are very bad. I'll try to post an example in english asap.

**Comment by snassimr** [comment]
Hmm ... Thanks you for your detailed explanation. Looks like input_ids messed up during evaluation. I am also will try to reproduce the issue for my case .

**Comment by pierpierpy** [comment]
came back with some news, i think the problem is that during training the model is trained on 4 bits (with BitsAndBytesConfig configuration)

**Comment by JHLiu7** [comment]
> > > i'm trying to decode with the tokenizer but it spits out some gibberish. half complete words and sentences. they make a bit of sense but they are pretty garbage. is there any solution to the problem? i'd like to have the real predictions to compute the rouge metric during training at each epoch.
> > 
> > 
> > Are you sure it's gibberish ? In the beginning it can look gibberish but improving .
> 
> The phrases produced during the training are "readable" but the words are misspelled. On the other hand, the predictions made using the model after the training seem to be just right, with no grammatical mistakes or misspellings. It is strange because i also tried to make the prediction during and after training with just one epoch. I assumed they should have been both wrong, but the results after training are still good, while the results during training are very bad. I'll try to post an example in english asap.
> 
> For this reason i think the problem is in the tokenizer for some misterious reasons, not in the training or the model itself.

It seems to me `preprocess_logits_for_metrics` is causing issues?  It's not entirely clear how it would interact with the decoding algorithm underpinning `generate` function. Or it may simply bypass `generate`.

**Comment by adamamer20** [comment]
As @cabosanlucas said, the solution provided by @isspek is "greedy search decoding", which is not optimal for consistency. See [here](https://huggingface.co/docs/transformers/generation_strategies) other strategies for generating text. Currently these strategies are only supported by `Seq2SeqTrainer`. Would be nice if support for `SFTTrainer` coule be added.

**Comment by NirmalManoj** [comment]
> As @cabosanlucas said, the solution provided by @isspek is "greedy search decoding", which is not optimal for consistency. See [here](https://huggingface.co/docs/transformers/generation_strategies) other strategies for generating text. Currently these strategies are only supported by `Seq2SeqTrainer`. Would be nice if support for `SFTTrainer` coule be added.

I am not sure if it's even greedy search decoding. Could you take a look at my comment above?

I plan to submit a PR for adding support to `SFTTrainer` in the future when I have time.

**Comment by isspek** [comment]
Hi everyone, thanks for checking this and finding out the issues. I hadnt throughly tested the method I tried. I am looking forward the correct version.

**Comment by JackCai1206** [comment]
Is this feature available now?

**Comment by adamamer20** [comment]
> Is this feature available now?

No

**Comment by mohammedkhalilia** [comment]
@adamamer20 you mentioned the workaround is using Seq2SeqTrainer with [Seq2SeqTrainingArguments(predict_with_generate=True, ...) ]

Is it possible to share your implementation if you have one?

**Comment by dabs9** [comment]
Adding my request for this feature to be added. Would be great to have SFTTrainer inherit from Seq2SeqTrainer.

**Comment by yznlp** [comment]
Linking my very rough code for a `SFTSeq2SeqTrainer` in case that's useful :) 

https://github.com/huggingface/trl/issues/1222#issuecomment-2498956319

**Comment by steveepreston** [comment]
+1

**Comment by kushwanth22** [comment]
+1

**Comment by steve2972** [comment]
+1

**Comment by cgr71ii** [comment]
+1

**Comment by cgr71ii** [comment]
Hi!

```python
def preprocess_logits_for_metrics(logits, labels):
    last_pred_token = logits.argmax(dim=-1)[:, -1].unsqueeze(0)
    generation = last_token.clone().detach()
    

    with torch.no_grad():
        while last_pred_token != tokenizer.eos_token_id or generation.size(1) < trainer.args.max_seq_length:
                logits = model.forward(torch.cat((labels, generation), dim=1)).logits
                last_pred_token = logits.argmax(dim=-1)[:, -1].unsqueeze(0)
                generation = torch.cat((generation, last_token), dim=1)
    
    return generation
```

Even if you provide this function provided by @adamamer20 to the `preprocess_logits_for_metrics` parameter in the `SFTTrainer`, you don't have access to the original input_ids, then you need to trust that your model has generated the same tokens that were provided as inputs. This might not be the case if you're finetuning a model that was trained using the `DataCollatorForCompletionOnlyLM` collator since loss for the prompt is not calculated.

**Comment by aaronkossler** [comment]
+1

**Comment by osuzdalev** [comment]
+1

**Comment by MauroCE** [comment]
+1

**Comment by maot-github** [comment]
+1

**Comment by steveepreston** [comment]
`predict_with_generate` for SFTTrainer still missing in 2025
This is so painful

**Comment by darkness8i8** [comment]
+1

**Comment by AndreiCComan** [comment]
+1

**Comment by ChangeFWorld** [comment]
Any update yet?

**Comment by steveepreston** [comment]
Why no one Re-Opens this?

**Comment by yurkoff-mv** [comment]
I'm also really looking forward to this option.

**Comment by steveepreston** [comment]
@wei-ann-Github Re-Open this issue please

**Comment by wei-ann-Github** [comment]
> [@wei-ann-Github](https://github.com/wei-ann-Github) Re-Open this issue please

@steveepreston , i don't have any permission more than you to open it.

**Comment by CodersAcademy006** [comment]
Hey @wei-ann-Github , I have added a fix for this, please review and provide me any feedback which improves. Thank You!!

**Comment by wei-ann-Github** [comment]
> Hey [@wei-ann-Github](https://github.com/wei-ann-Github) , I have added a fix for this, please review and provide me any feedback which improves. Thank You!!

Hi @CodersAcademy006 , my apologies,  i am not familiar with the rules of open source projects. I am a member of the public just like you. I do not have admin rights.  I am not sure if I am the right person to review.

**Comment by wei-ann-Github** [comment]
> > Hey [@wei-ann-Github](https://github.com/wei-ann-Github) , I have added a fix for this, please review and provide me any feedback which improves. Thank You!!
> 
> Hi [@CodersAcademy006](https://github.com/CodersAcademy006) , my apologies, i am not familiar with the rules of open source projects. I am a member of the public just like you. I do not have admin rights. I am not sure if I am the right person to review.

@murilo-cunha , can you advice @CodersAcademy006 please?

**Comment by murilo-cunha** [comment]
> > > Hey [@wei-ann-Github](https://github.com/wei-ann-Github) , I have added a fix for this, please review and provide me any feedback which improves. Thank You!!
> > 
> > 
> > Hi [@CodersAcademy006](https://github.com/CodersAcademy006) , my apologies, i am not familiar with the rules of open source projects. I am a member of the public just like you. I do not have admin rights. I am not sure if I am the right person to review.
> 
> [@murilo-cunha](https://github.com/murilo-cunha) , can you advice [@CodersAcademy006](https://github.com/CodersAcademy006) please?

Hi I just did a small contribution to the CONTRIBUTING.md
The other contribution is still an open PR
So I don't have a lot of insights to give you on this project in particular.

I guess it's ok to review and give comments on the other PR @wei-ann-Github, but you won't be able to merge anything
